### PR TITLE
plugins: add support for ndpi-like plugins - v1

### PR DESCRIPTION
**Pull request by jasonish**
Non-draft version of nDPI plugin support.

Previous PR: https://github.com/OISF/suricata/pull/11944

Includes:
- flow lifecycle callbacks
- thread init callback
- eve output callback

### CODE REVIEW

File: src/flow-callbacks.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * 02110-1301, USA.
+ */
+
+#ifndef SURICATA_FLOW_CALLBACKS_H
+#define SURICATA_FLOW_CALLBACKS_H
+
+#include "suricata-common.h"
+#include "flow.h"
+
+typedef void (*SCFlowInitCallbackFn)(ThreadVars *tv, Flow *f, const Packet *p, void *user);
```
Comment (victorjulien) [inline src/flow-callbacks.h:24]: Think we should document these typedefs as well as the place where the callback itself is registered? When is it called, with what args, etc.

File: src/output-eve.c [hunk 2]
```diff
@@ -15,11 +15,50 @@
...
+    EveUserCallbackFn fn;
+    void *user;
+    struct EveUserCallback_ *next;
+} EveUserCallback;
+
+static EveUserCallback *eve_user_callbacks = NULL;
 
 static TAILQ_HEAD(, SCEveFileType_) output_types = TAILQ_HEAD_INITIALIZER(output_types);
 
+bool EveRegisterCallback(EveUserCallbackFn fn, void *user)
```
Comment (victorjulien) [inline src/output-eve.c:33]: Should the SC prefix be added here as well?

File: src/output-json.c [hunk 3]
```diff
@@ -24,6 +24,7 @@
  *
  */
 
+#include "output-eve.h"
```
Comment (victorjulien) [inline src/output-json.c:27]: typically suricata-common is the first include

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 23113

File: src/flow-manager.c [hunk 4]
```diff
@@ -1058,7 +1059,7 @@ static void Recycler(ThreadVars *tv, FlowRecyclerThreadData *ftd, Flow *f)
         StatsDecr(tv, ftd->counter_tcp_active_sessions);
     }
     StatsDecr(tv, ftd->counter_flow_active);
-
+    SCFlowRunFinishCallbacks(tv, f);
```
Comment (victorjulien) [inline src/flow-manager.c:1062]: need this also in `FlowGetUsedFlow` I think

**Review by victorjulien — CHANGES_REQUESTED** [review]
Looks good. See inline comments on the naming prefix.

Also, wonder if these new files should start in a new dir structure (remember that one? :))

**Comment by jasonish** [comment]
> Also, wonder if these new files should start in a new dir structure (remember that one? :))

Not yet, I don't think.  We have to decide if we should do the re-org to a new directory as well, to look and feel more like a first class library: https://github.com/OISF/suricata/pull/10421

**Pull request closed without merging**